### PR TITLE
Add data attribute to papermill parameter cells

### PR DESCRIPTION
**Pull request by amad-a**
## References

#8298 Add Data Attribute to Parameter Cells

## Code changes

- Create a new case in the Metadata change handler that checks for a 'tags' key in the metadata. Then, this case calls a new function checkParameters() which checks if the metadata has a 'parameters' tag, and sets or removes a 'data-parameter-cell' attribute in the cell's CSS.

## User-facing changes

None

## Backwards-incompatible changes

None I am aware of

### CODE REVIEW

File: packages/cells/src/widget.ts [hunk 1]
```diff
@@ -327,6 +326,17 @@ export class Cell extends Widget {
     this.readOnly = this.model.metadata.get('editable') === false;
   }
 
+  /**
+   * Checks if cell has the papermill 'parameters' tag, and sets a CSS attribute accordingly.
+   */
+  checkParameters() {
```
Comment (blink1073) [inline packages/cells/src/widget.ts:332]: I think we should make this method private.

**Comment by jasongrout** [comment]
Doesn't it seem a little awkward to hardcode knowledge of papermill in JupyterLab? Are we going to hardcode knowledge of other community tools into JupyterLab too?

It would make more sense to me if we had a Jupyter standard for such cell metadata, then wrote our code to conform to the standard. Or if we made a more generic mechanism for annotating cells that a separate papermill plugin could use.

**Comment by blink1073** [comment]
I'd argue papermill is widely used enough that acknowledging its existence in core makes sense in some fashion.  This was a low-hanging fruit way to accomplish that.   I'm wary of adding too many ambitious APIs around the notebook model when we have a large refactor coming to the notebook models.

**Comment by jasongrout** [comment]
> I'd argue papermill is widely used enough that acknowledging its existence in core makes sense in some fashion.

I understand. I'd argue that the appropriate way to acknowledge it in core tools is to codify it in the notebook standard metadata, the same as we did with the execution timing, cell hiding, etc.

**Comment by blink1073** [comment]
It is using a standard, "tag".  Perhaps we should add `data-tag-foo` attributes for all tags?

**Comment by jasongrout** [comment]
>  Perhaps we should add `data-tag-foo` attributes for all tags?

Sure, I'd support annotating cells in a generic way based on the tags the cell has.

However, as the code is written, it seems like it is querying for the top-level metadata key "parameters"? That seems to be the papermill-specific part.

In general, the problem with building support for specific things in the core tools, but not having it in the standard, is that then it becomes a de facto standard which diverges from the real notebook standard, undercutting the benefits of standardization in the first place.

**Comment by blink1073** [comment]
> However, as the code is written, it seems like it is querying for the top-level metadata key "parameters"? That seems to be the papermill-specific part.

Whoops, I missed that part.  

> In general, the problem with building support for a specific thing in the core tools, but not having it in the standard, is that then it becomes a de facto standard which diverges from the real notebook standard, undercutting the benefits of standardization in the first place.

Fair enough

**Comment by blink1073** [comment]
@amad-a, apologies for the side-bar on your PR.

The recommended changes are to:

- Clear any existing `data-tag-<foo>` parameters on `this.node` using [`this.node.dataset`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/dataset)
- Add `data-tag-foo` attributes for all of the tags in `metadata.tags` (check for `metadata.has('tags')` first)
- Rename `checkParameters()` -> `private _handleTags()`

**Comment by jasongrout** [comment]
So papermill parameters *are* just normal cell tags? In other words, in the `metadata.tags` key?

**Comment by jasongrout** [comment]
Looking at https://papermill.readthedocs.io/en/latest/usage-parameterize.html#lab, I think I understand now. Papermill tags a cell with the "parameters" tag, which is just a generic tag, so we actually won't have anything about "parameters" in here at all. Just labeling nodes with `data-tag-foo` for every tag "foo". Sure, that sounds great.

**Comment by jasongrout** [comment]
Just curious, for those that will presumably be styling or creating shortcuts for these cells - what guarantees are we making about the DOM structure of nodes we are tagging with these tags? If we change the DOM structure of cells, what will it break and are we supporting it as a public API?

**Comment by blink1073** [comment]
A good reason to preserve class names where possible.  If we change class names or hierarchy, it is a breaking API change and any extensions targeting CSS attributes would need to be updated.

**Comment by jasongrout** [comment]
I think at least most of the time, we do *not* consider the DOM structure a public API. We've never clarified exactly what parts are there, so things are vague. Same with class names - we've talked about *some* class names being public API, but we've not clarified which ones.

**Comment by blink1073** [comment]
Many of the CSS selectors we are using, including for keyboard shortcuts, use parent -> child relationships of CSS classes and data attributes.

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/styling-cells-through-metadata/4978/9

**Comment by blink1073** [comment]
Closing in favor of #8627 due to lack of response.

**Pull request closed without merging**